builtin/method_str.py

```diff
@@ -130,3 +135,148 @@ def Call(self, rd):
             return value.Null
 
         return RegexMatch(string, indices, capture)
+
+
+class Replace(vm._Callable):
+
+    def __init__(self, mem, expr_ev):
+        # type: (state.Mem, expr_eval.ExprEvaluator) -> None
+        self.mem = mem
+        self.expr_ev = expr_ev
+
+    def EvalSubstExpr(self, expr, blame_loc):
+        # type: (value.Expr, loc_t) -> str
+        res = self.expr_ev.EvalExpr(expr.e, blame_loc)
+        if res.tag() == value_e.Str:
+            return cast(value.Str, res).s
+
+        raise error.TypeErr(res, "expected expr to eval to a string",
+                            blame_loc)
+
+    def Call(self, rd):
+        # type: (typed_args.Reader) -> value_t
+        """
+        s => replace(string_val, subst_str, count=-1)
+        s => replace(string_val, subst_expr, count=-1)
+        s => replace(eggex_val, subst_str, count=-1)
+        s => replace(eggex_val, subst_expr, count=-1)
+        """
+        string = rd.PosStr()
+
+        string_val = None  # type: value.Str
+        eggex_val = None  # type: value.Eggex
+        subst_str = None  # type: value.Str
+        subst_expr = None  # type: value.Expr
+
+        pattern = rd.PosValue()
+        with tagswitch(pattern) as case:
+            if case(value_e.Eggex):
+                # HACK: mycpp will otherwise generate:
+                #  value::Eggex* eggex_val ...
+                eggex_val_ = cast(value.Eggex, pattern)
```

Comment:
I think this is the reason for the awkward UP pattern, e.g. I think you can use `UP_pattern` and `UP_subst` to be consistent with the rest of the codebase


There is a Zulip thread about that, let me find this